File: apps/web/lib/ct-warehouse-labels.ts
/** Etiquetas de almacenes CT para UI (paridad con API `ct-warehouses.ts`). */

export type CtWarehouseOption = { code: string; label: string; city?: string };

export type StockByWarehouseRow = { code: string; qty: number; label?: string; city?: string };

const CATALOG_CODE_LABELS: Record<string, string> = {
  PUE: "Puebla",
  MTY: "Monterrey",
  GDL: "Guadalajara",
  CDMX: "CDMX",
  DFA: "CDMX",
  TPC: "Toluca",
  HMO: "Hermosillo",
  CHI: "Chihuahua",
  LEO: "León",
  "14A": "Puebla",
  "35A": "Monterrey",
  "46A": "Guadalajara",
  "13A": "CDMX",
  "01A": "Hermosillo",
  "03A": "Chihuahua",
  "07A": "León",
};

export function warehouseLabel(code: string): string {
  const key = String(code || "").trim().toUpperCase();
  if (!key) return "—";
  return CATALOG_CODE_LABELS[key] || key;
}

/** Códigos de catálogo que corresponden a una ciudad (Puebla, Monterrey, …). */
export function codesForCity(city: string): string[] {
  const c = city.toLowerCase();
  if (c.includes("puebla")) return ["PUE", "14A"];
  if (c.includes("monterrey")) return ["MTY", "35A"];
  if (c.includes("guadalajara")) return ["GDL", "46A"];
  if (c.includes("cdmx") || c.includes("mexico")) return ["CDMX", "DFA", "13A"];
  if (c.includes("hermosillo")) return ["HMO", "01A"];
  return [];
}

export function stockAtPreferred(
  rows: StockByWarehouseRow[] | undefined,
  preferredCodes: string[],
): number {
  if (!rows?.length || !preferredCodes.length) return 0;
  const set = new Set(preferredCodes.map((c) => c.toUpperCase()));
  return rows.reduce((sum, r) => (set.has(r.code.toUpperCase()) ? sum + r.qty : sum), 0);
}

export function formatStockByWarehouse(
  rows: StockByWarehouseRow[] | undefined,
  opts?: { max?: number; preferredCodes?: string[] },
): string {
  if (!rows?.length) return "";
  const max = opts?.max ?? 4;
  const pref = new Set((opts?.preferredCodes ?? []).map((c) => c.toUpperCase()));
  const sorted = [...rows].sort((a, b) => {
    const ap = pref.has(a.code.toUpperCase()) ? 1 : 0;
    const bp = pref.has(b.code.toUpperCase()) ? 1 : 0;
    if (bp !== ap) return bp - ap;
    return b.qty - a.qty;
  });
  return sorted
    .slice(0, max)
    .map((r) => `${r.label || warehouseLabel(r.code)} ${r.qty}`)
    .join(" · ");
}

File: apps/mobile-native/android/app/src/main/java/mx/nexara/mobile/nativeapp/data/crm/CtWarehouseLabels.kt
package mx.nexara.mobile.nativeapp.data.crm

import mx.nexara.mobile.nativeapp.data.api.StockByWarehouseDto

private val CATALOG_CODE_LABELS = mapOf(
    "PUE" to "Puebla",
    "MTY" to "Monterrey",
    "GDL" to "Guadalajara",
    "CDMX" to "CDMX",
    "DFA" to "CDMX",
    "TPC" to "Toluca",
    "HMO" to "Hermosillo",
    "CHI" to "Chihuahua",
    "LEO" to "León",
    "14A" to "Puebla",
    "35A" to "Monterrey",
    "46A" to "Guadalajara",
    "13A" to "CDMX",
    "01A" to "Hermosillo",
    "03A" to "Chihuahua",
    "07A" to "León",
)

fun warehouseLabel(code: String): String {
    val key = code.trim().uppercase()
    return CATALOG_CODE_LABELS[key] ?: key.ifBlank { "—" }
}

fun warehouseRowLabel(row: StockByWarehouseDto): String {
    return row.label?.takeIf { it.isNotBlank() } ?: warehouseLabel(row.code)
}

/** Códigos de catálogo que corresponden a una ciudad (Puebla, Monterrey, …). */
fun codesForCity(city: String): List<String> {
    val c = city.lowercase()
    return when {
        "puebla" in c -> listOf("PUE", "14A")
        "monterrey" in c -> listOf("MTY", "35A")
        "guadalajara" in c -> listOf("GDL", "46A")
        "cdmx" in c || "mexico" in c -> listOf("CDMX", "DFA", "13A")
        "hermosillo" in c -> listOf("HMO", "01A")
        else -> emptyList()
    }
}

fun stockAtPreferred(rows: List<StockByWarehouseDto>?, preferredCodes: List<String>): Int {
    if (rows.isNullOrEmpty() || preferredCodes.isEmpty()) return 0
    val set = preferredCodes.map { it.uppercase() }.toSet()
    return rows.sumOf { if (set.contains(it.code.uppercase())) it.qty else 0 }
}

fun sortedWarehouseRows(
    rows: List<StockByWarehouseDto>?,
    preferredCodes: List<String> = emptyList(),
    max: Int = 5,
): List<StockByWarehouseDto> {
    if (rows.isNullOrEmpty()) return emptyList()
    val pref = preferredCodes.map { it.uppercase() }.toSet()
    return rows
        .filter { it.qty > 0 }
        .sortedWith(
            compareByDescending<StockByWarehouseDto> { pref.contains(it.code.uppercase()) }
                .thenByDescending { it.qty },
        )
        .take(max)
}

fun formatStockByWarehouse(
    rows: List<StockByWarehouseDto>?,
    max: Int = 4,
    preferredCodes: List<String> = emptyList(),
): String {
    return sortedWarehouseRows(rows, preferredCodes, max)
        .joinToString(" · ") { "${warehouseRowLabel(it)} ${it.qty}" }
}

fun formatLeadTimeDays(days: Int): String = when {
    days <= 0 -> ""
    days <= 1 -> "Inmediata"
    else -> "$days días"
}

fun hasPromotion(promociones: List<Map<String, Any?>>?): Boolean = !promociones.isNullOrEmpty()

fun badgeLabelEs(badge: String): String = when (badge.uppercase()) {
    "RECOMMENDED" -> "Top"
    "BEST_PRICE" -> "Mejor precio"
    "BEST_STOCK" -> "Más stock"
    "FASTEST" -> "Más rápido"
    "BEST_MARGIN" -> "Mejor margen"
    "SUBSTITUTE" -> "Sustituto"
    else -> badge
}
